### PR TITLE
Add HUD scaling config parameter

### LINKED ISSUE #609: Add HUD scaling for high resolutions
In the example screenshots HUD scaled x2 for 1920x1080 (made in paint.net)

rwgame 2018-08-28 03-36-28-35

x2

### PR DESCRIPTION
**Pull request by qkolj**
This pull request adds hud_scale parameter to game config file. I needed to change both DrawUI and MapRenderer because while DrawUI handles most of the HUD drawing parameters, parameters for drawing radar blips are actually all handled inside MapRenderer class. If someone has an idea how to refactor code to make HUD scaling handled more centrally, please share. Closes #609 

Ingame screenshots:

hud_scale=1 on 1920x1080
screenshot_20181030_132022

hud_scale=1.5 on 1920x1080
screenshot_20181030_132053

hud_scale=1.75 on 1920x1080
screenshot_20181030_132124

hud_scale=2 on 1920x1080
screenshot_20181030_132150

### CODE REVIEW

**Comment by qkolj** [comment]
This is a bit ugly:
https://github.com/rwengine/openrw/blob/e767f2c68628cafd0b8c251a11ccd23df17ba2cb/rwgame/DrawUI.cpp#L38-L50
I assume there is a point to keeping the original values unchanged, but if there isn't, I can change this to remove unnecessary code duplication.

File: rwgame/DrawUI.cpp [hunk 1]
```diff
@@ -35,15 +35,29 @@ constexpr float ui_mapSize = 150.f;
 constexpr float ui_worldSizeMin = 200.f;
 constexpr float ui_worldSizeMax = 300.f;
```
Comment (ShFil119) [inline rwgame/DrawUI.cpp]: nit: Inconsistency of naming. Should be used camel case.

File: rwengine/src/render/MapRenderer.cpp [hunk 2]
```diff
@@ -289,3 +289,9 @@ void MapRenderer::drawBlip(const glm::vec2& coord, const glm::mat4& view,
     renderer->setUniform(rectProg.get(), "colour", glm::vec4(0.0f, 0.0f, 0.0f, 1.0f));
     glDrawArrays(GL_LINE_LOOP, 0, 4);
 }
+
+void MapRenderer::setHUDScale(const float scale) {
```
Comment (ShFil119) [inline rwengine/src/render/MapRenderer.cpp]: `applyHUDScale` or `scaleHUD` should be better. Now it sounds like normal setter.

File: rwgame/DrawUI.cpp [hunk 3]
```diff
@@ -35,15 +35,29 @@ constexpr float ui_mapSize = 150.f;
...
 
+static float hudScale = 1.f;
```
Comment (ShFil119) [inline rwgame/DrawUI.cpp]: You shouldn't use global variables which are not `const`. Better insert these into class.

File: rwgame/DrawUI.cpp [hunk 4]
```diff
@@ -330,3 +344,19 @@ void drawOnScreenText(GameWorld* world, GameRenderer* renderer) {
         }
     }
 }
+
+void setHUDScale(const float scale) {
```
Comment (ShFil119) [inline rwgame/DrawUI.cpp]: Like earlier.

File: rwgame/GameConfig.cpp [hunk 5]
```diff
@@ -129,6 +129,22 @@ struct IntTranslator {
     }
 };
 
+struct FloatTranslator {
```
Comment (ShFil119) [inline rwgame/GameConfig.cpp]: nit: structures of translators should be refactored.

File: rwgame/RWGame.cpp [hunk 6]
```diff
@@ -108,6 +108,9 @@ RWGame::RWGame(Logger& log, int argc, char* argv[])
         }
     });
 
+    setHUDScale(config.getHUDScale());
```
Comment (ShFil119) [inline rwgame/RWGame.cpp]: Can code of these lines be located closer to comment `Set up text renderer`? For example after line 77 or 82.

**Comment by ShFil119** [comment]
1) You can prepare some wrapper for default settings, it should look a bit nicer.
2) "If someone has an idea how to refactor code to make HUD scaling handled more centrally, please share." I think it's problem of hardcodes. Positions should be calculated.

**Comment by qkolj** [comment]
I made most of your suggested changes. I didn't refactor translators in the GameConfig class because I think that is kind of unrelated to this PR and should be done in a separate PR.

**Comment by ShFil119** [comment]
After talking with danhendron, I think actually it's in acceptable state. Do you want to do more in this pull?

**Comment by ShFil119** [comment]
:+1: for C++ class and tests. :)

File: rwgame/RWGame.cpp [hunk 7]
```diff
@@ -75,6 +74,10 @@ RWGame::RWGame(Logger& log, int argc, char* argv[])
     renderer.text.setFontTexture(FONT_PRICEDOWN, "font1");
     renderer.text.setFontTexture(FONT_ARIAL, "font2");
 
+    hudDrawer = HUDDrawer();
```
Comment (ShFil119) [inline rwgame/RWGame.cpp]: You can avoid this line, if you convert 35th line of `RWGame.hpp` to `HUDDrawer hudDrawer{};`.

File: rwgame/HUDDrawer.hpp [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        float uiMapSize = 150.f;
+        float uiWorldSizeMin = 200.f;
+        float uiWorldSizeMax = 300.f;
+    };
+
+    HUDDrawer() = default;
+    ~HUDDrawer() = default;
+
+    void applyHUDScale(const float scale);
+    HUDParameters getHUDParameters();
```
Comment (ShFil119) [inline rwgame/HUDDrawer.hpp]: `const`?

File: rwgame/HUDDrawer.cpp [hunk 9]
```diff
@@ -330,3 +305,23 @@ void drawOnScreenText(GameWorld* world, GameRenderer* renderer) {
         }
     }
 }
+
+void HUDDrawer::applyHUDScale(const float scale) {
+	hudParameters.hudScale = scale;
```
Comment (ShFil119) [inline rwgame/HUDDrawer.cpp]: indention

**Comment by husho** [comment]
Scaling does not seem to work properly judging by your own screenshots

**Comment by qkolj** [comment]
@husho Those were old screenshots from the first version. It has been fixed since, but I forgot to update screenshots. I updated OP with screenshots from the final version.

**Comment by ShFil119** [comment]
Thanks. :)

**Pull request merged**